Fix #2728: Add AlgorithmRegistry and test class.

anmolshkl opened the pull request.

Andimeo commented on core/domain/classifier/algorithm_registry.py:
```diff
@@ -0,0 +1,96 @@
+
+    @classmethod
+    def get_all_classifier_ids(cls):
+        """Get a list of all algorithm ids.
```
`Get` -> `Gets`

anmolshkl replied:
Done.

Andimeo commented on core/domain/classifier/algorithm_registry.py:
```diff
@@ -0,0 +1,96 @@
+
+    @classmethod
+    def _refresh(cls):
+        """Refresh the dict mapping algorithm ids to instances."""
```
`Refresh` -> `Refreshes` and ditto elsewhere

anmolshkl replied:
Done.

Andimeo commented on core/domain/classifier/algorithm_registry.py:
```diff
@@ -0,0 +1,96 @@
+        # registry.
+        for classifier_id in all_classifier_ids:
+            my_class = cls.custom_import(''.join([feconf.CLASSIFIERS_PKG, '.',
+                                                  classifier_id, '.',
```
Oh, I see why you rename `lda_string_classifier` to `LDAStringClassifier`. Can we add something in `core.domain.classifier.__init__.py` like `import lda_string_classifier as LDAClassifier` to avoid the renaming?

anmolshkl replied:
I have followed this convention from interaction_registry.py. If you notice then you'll find that it imports module on the basis of interaction IDs itself (from extensions/interactions/...).

seanlip replied:
On balance I prefer not editing `__init__.py`. I think LDAStringClassifier is probably OK in this case.

BenHenning replied:
I'm not sure I understand why the rename was necessary. Can't we join on ``_`` to create the class name?

seanlip replied:
I think the idea is to follow the existing pattern in extensions/interactions (and our other registries), where the file names are capitalized to match the item IDs (like MultipleChoiceInput). On the whole, I think it's worth keeping the consistency here, rather than having two different conventions for different registries. What do you think, @BenHenning?

BenHenning replied:
If that's the case, then I suggest we move ``LDAClassifier`` to the extensions folder. If it's following naming conventions in extensions, it should probably be one (I think it also makes sense to be an extension given the nature of these classifiers).

seanlip replied:
I'd be OK with that. @anmolshkl does this sound good?

anmolshkl replied:
What does 'extension' actually means here? As in, what's the use of extension folder?

seanlip replied:
The folders in 'extensions' currently contain things that are accessed via a registry, such as interactions, objects, etc. The idea is that it should be easy to add a new extension by adding a new subdirectory, file or class to the relevant extensions file/folder, and it will behave the same way as the other extensions. E.g., if I wanted to add a new interaction, I just need to add a new folder to the extensions/interactions directory and not touch the rest of the codebase. Arguably the same would be true of classifiers.

anmolshkl replied:
Done. But you know, now the things look very confusing :p. 
It's like from "..LDAStringClassifier import LDAStringClassifier...." and then "LDAStringClassifier.LDAStringClassifier()". Phew!

seanlip replied:
That shouldn't happen, because now that you have a registry, you shouldn't be doing a direct import. Instead, the only top-level import is for the registry, and the classifier class should be obtained from by a get() from the registry.

anmolshkl replied:
Oh right, I didn't use the registry till now because there is still some issues with it. This is just a temporary issue in that case :D

Andimeo commented on core/domain/classifier/algorithm_registry.py:
```diff
@@ -0,0 +1,96 @@
+
+        Returns:
+            An instance of the classifier.
+        """
```
The KeyError should be portrayed in another section.

anmolshkl replied:
Ah right! Thanks for spotting. I'll have to update that in interaction registry as well.

Andimeo commented on core/domain/classifier/algorithm_registry_test.py:
```diff
@@ -0,0 +1,42 @@
+
+from core.domain.classifier import algorithm_registry
+from core.tests import test_utils
+from core.domain.classifier.LDAStringClassifier import LDAStringClassifier
```
Alphabetic order.

anmolshkl replied:
Done.

Andimeo commented on core/domain/classifier/algorithm_registry_test.py:
```diff
@@ -0,0 +1,42 @@
+
+    def test_get_all_classifiers(self):
+        classifier_instances = algorithm_registry.Registry.get_all_classifiers()  # pylint: disable=line-too-long
+        self.assertEquals(len(classifier_instances), 1)
```
Can we retrieve the expected results from `feconf`? 
This is a test case for `AlgorithmRegistry`, but every time we modify `feconf` it will be broke.

anmolshkl replied:
Done.

seanlip commented on core/controllers/reader.py:
```diff
@@ -21,7 +21,7 @@
 
 from core.controllers import base
-from core.domain.classifier import lda_string_classifier
+from core.domain.classifier import LDAStringClassifier
```
A general module like reader.py should not inherit a specific module like LDAStringClassifier. Instead, use the classifier registry (which is much more easily extensible) to get the classifier class. This is similar to what we do for interactions.

anmolshkl replied:
@seanlip I haven't really touched reader.py other than updating the module name. Actually, reader.py has to be updated to use the classifier_registry. I have left these changes for the next phase (since I've planned to revamp the classification specific functions in reader.py anyway).

For now, I have only updated the names so that the things do not break.

seanlip replied:
Ah. I think in this PR you should actually update it to use the registry, since you've just introduced the registry. That's kind of the point of the PR, and is a good test of whether the whole thing works properly.

anmolshkl replied:
OK, I'll do that. Thanks!

seanlip commented on core/domain/classifier/LDAStringClassifierTest.py:
```diff
@@ -14,7 +14,7 @@
 # limitations under the License.
 
-from core.domain.classifier.lda_string_classifier import LDAStringClassifier
+from core.domain.classifier.LDAStringClassifier import LDAStringClassifier
```
Nope -- from core.domain.classifier import modulename. We always import entire files (and reference a class by modulename.ClassName); we don't import specific classes.

anmolshkl replied:
Oh, thanks for spotting that. Good that we found this after having missed it in Yang Lu's PR.

anmolshkl replied:
Done.

seanlip commented on core/domain/classifier/algorithm_registry.py:
```diff
@@ -0,0 +1,104 @@
+class Registry(object):
+    """Registry of all classification algorithms/classifiers."""
+
+    # Dict mapping algorithm ids to instances of the classification algorithms.
```
"instances of the algorithm classes", perhaps. I think you want to convey the fact that the value is a 'class'; if you tell me it's an 'algorithm' I'm not quite sure what you mean.

anmolshkl replied:
Yep, sounds better. Thanks!

BenHenning replied:
Did you make this change? The string seems to still say ``classification algorithms``.

anmolshkl replied:
I was stuck on something so I haven't pushed anything. But yea, I have made this particular change.

anmolshkl replied:
Done.

seanlip commented on core/domain/classifier/algorithm_registry.py:
```diff
@@ -0,0 +1,104 @@
+    _classification_algorithms = {}
+
+    @classmethod
+    def custom_import(cls, path):
```
Hm, why do we need this? Can't we use the existing pattern in, e.g., core.domain.interaction_registry? I think there's something to be said for maintaining consistency...

anmolshkl replied:
Ah I was going to bring it up. Actually, I followed exactly the same approach from interaction_registry.py but weirdly it didn't work for me. `for loader, name, _ in pkgutil.iter_modules(path=classifier_paths)` couldn't find anything with `classifier_paths` set to ['core/domain/classifier/LDAStringClassifier']. That was the reason for shifting to the above approach. 

That said, even I think that consistency is an issue here. I'll try again with the alternative approach and get back to you if it's not working again.

anmolshkl replied:
Done. I've removed this method in the new commit.

seanlip commented on core/domain/classifier/algorithm_registry.py:
```diff
@@ -0,0 +1,104 @@
+        Returns:
+            An instance of the classifier.
+        """
+
```
Remove newline.

anmolshkl replied:
Done.

seanlip commented on core/domain/classifier/algorithm_registry_test.py:
```diff
@@ -0,0 +1,49 @@
+            classifier_class['algorithm_ids']
+            for classifier_class in feconf.ALLOWED_CLASSIFIER_CLASSES
+        ])))
+        classifier_instances = algorithm_registry.Registry.get_all_classifiers()  # pylint: disable=line-too-long
```
No need to disable line-too-long. Instead:
```
    classifier_instances = (
        algorithm_registry.Registry.get_all_classifiers())
```

Ditto below.

anmolshkl replied:
Done.

seanlip commented on core/domain/classifier/algorithm_registry_test.py:
```diff
@@ -0,0 +1,49 @@
+    def test_get_classifier_by_id(self):
+        classifier_id = 'LDAStringClassifier'
+        classifier_instance = algorithm_registry.Registry.get_classifier_by_id(classifier_id)  # pylint: disable=line-too-long
+        self.assertIsInstance(classifier_instance, LDAStringClassifier,
```
FYI, better to break at '(', in general. Then you don't have to cram everything on the right.

anmolshkl replied:
Done.

seanlip commented on feconf.py:
```diff
@@ -47,6 +47,7 @@
 else:
     raise Exception('Invalid platform: expected one of [\'gae\']')
 
+CLASSIFIERS_PKG = 'core.domain.classifier'
```
Ditto -- maybe follow the existing patterns rather than create a new one?

anmolshkl replied:
Will remove it and would use CLASSIFIERS_DIR instead once the alternative approach works.

BenHenning replied:
Will that work happen in this PR?

anmolshkl replied:
Right now I'm trying to make the interaction_registry's approach of importing modules using "pkgutils.itermodules ...' work (I couldn't make it work initially and so I had changed the approach). Btw, I'm almost done and these changes will be reflected in my next commit.

BenHenning commented on core/domain/classifier/algorithm_registry.py:
```diff
@@ -0,0 +1,104 @@
+    """Registry of all classification algorithms/classifiers."""
+
+    # Dict mapping algorithm ids to instances of the classification algorithms.
+    _classification_algorithms = {}
```
Perhaps ``_classifier_classes`` would be a more descriptive name. It helps address Sean's early point about communicating this dict contains classes.

anmolshkl replied:
Thanks for pointing out, I've changed this already. I'll notify you when I have pushed the commits.

anmolshkl replied:
Done. Renamed it to `_classifier_instances` instead.

BenHenning commented on core/domain/classifier/algorithm_registry.py:
```diff
@@ -0,0 +1,104 @@
+            A list of instances of all the classification algorithms.
+        """
+
+        if len(cls._classification_algorithms) == 0:
```
More pythonic:

```python
if not cls._classification_algorithms:
```

anmolshkl replied:
Though as much but then stuck to interaction_registry's pattern. Will update it.

seanlip replied:
OK to update this elsewhere too, if you like.

anmolshkl replied:
Done!

BenHenning commented on core/domain/classifier/algorithm_registry_test.py:
```diff
@@ -0,0 +1,49 @@
+        self.assertItemsEqual(expected_algorithm_ids, algorithm_ids)
+
+    def test_get_all_classifiers(self):
+        num_classifiers = len(list(itertools.chain(*[
```
Isn't this duplicating logic from the registry class? I wonder if we could instead just assert how many classifiers there are without performing this logic. We'd need to update this test as we add/remove classifiers, but that's probably fine.

anmolshkl replied:
Umm, the earlier approach was to actually just assert how many classifiers are there. But @Andimeo raised the concern that it'd break things every time we add a new classifier (I agree!). I'm not sure what's better here. @seanlip any advice?

seanlip replied:
I think it is fine to just assert how many classifiers there are. We won't be adding new classifiers that often, and I'm fine with this value having to be changed manually each time one is added.

anmolshkl replied:
Cool. Done!

BenHenning commented on core/domain/classifier/algorithm_registry_test.py:
```diff
@@ -0,0 +1,49 @@
+        classifier_id = 'LDAStringClassifier'
+        classifier_instance = algorithm_registry.Registry.get_classifier_by_id(classifier_id)  # pylint: disable=line-too-long
+        self.assertIsInstance(classifier_instance, LDAStringClassifier,
+                              msg='The object is not an instance of '
```
Is the msg necessary? Won't the error upon failure already indicate this to some extent?

anmolshkl replied:
Yes, it would indicate to some extent. I was just trying to follow the [pattern](https://github.com/oppia/oppia/blob/develop/extensions/objects/models/objects_test.py#L422) here.

BenHenning commented on feconf.py:
```diff
@@ -348,6 +349,14 @@ def get_empty_ratings():
 }
 
+# Ids of the allowed classification algorithm classes.
+ALLOWED_CLASSIFIER_CLASSES = [{
```
Why 'allowed'? Why not just: ``ANSWER_CLASSIFICATION_CLASSES``?

anmolshkl replied:
Just like 'ALLOWED_INTERACTION_CATEGORIES'. But will change it if it doesn't make sense.

seanlip replied:
The reason we had 'ALLOWED' for interactions was because some interactions were being developed as test interactions and we didn't want them to be user-visible while they were being developed. Not sure if this is going to apply to classifiers, so I defer to y'all here.

anmolshkl replied:
Done. Renamed it to `ANSWER_CLASSIFICATION_CLASSES`.

BenHenning commented on feconf.py:
```diff
@@ -348,6 +349,14 @@ def get_empty_ratings():
     },
 }
 
+# Ids of the allowed classification algorithm classes.
```
Can you comment a bit about how the dicts need to be structured in this list? It's not immediately obvious what name/algorithm_ids are. If you can refer to somewhere else in code that explains it, that's fine, too. Either way, it should be easier to understand what this list contains and how to change it.

anmolshkl replied:
Ok, sure! I'll do that.

anmolshkl replied:
Done!

seanlip commented:
FWIW, I like `classifier_registry` -- that seems like a good suggestion to me. @anmolshkl it's never the case that a classifier corresponds to multiple algorithms, right?

anmolshkl commented:
@seanlip Um yes, even I was thinking of changing it to classifier_registry. No, a particular type of classifier will (and should not, ideally) correspond to multiple algorithms.

anmolshkl commented:
@seanlip Can you PTAL? The third test in the classifier_registry_test.py is failing due to some reason and I can't figure it out. Can you please take a look and let me know if you find something. Meanwhile, I'm also investigating into it.

anmolshkl commented:
@seanlip @BenHenning Done! I have made all the changes. All the back-end tests are passing as well. PTAL, thanks!

seanlip commented on core/controllers/reader.py:
```diff
@@ -21,7 +21,7 @@
 
 from core.controllers import base
-from core.domain.classifier import lda_string_classifier
+from core.domain.classifier import classifier_registry
```
I suggest putting classifier_registry in core.domain (since we're moving all the other stuff to extensions) and getting rid of the classifiers folder.

anmolshkl replied:
Ok, sounds reasonable.

anmolshkl replied:
Done!

seanlip commented on feconf.py:
```diff
@@ -68,6 +69,9 @@
     os.getcwd(), 'extensions', 'interactions', 'rules.json')
 
+# The ID of the classifier to be used as a default string classifier.
+DEFAULT_STRING_CLASSIFIER = 'LDAStringClassifier'
```
CLASSIFIER_ID_LDA_STRING_CLASSIFIER (or CLASSIFIER_ID_LDA). Basically, start all these things with CLASSIFIER_ID_.

anmolshkl replied:
Ok, sure.

anmolshkl replied:
Btw, tomorrow this constant might point to a different classifier. Would it be wise to name it as CLASSIFIER_ID_LDA_STRING_CLASSIFIER? I think DEFAULT_STRING_CLASSIFIER (or a similar name) makes more sense.

seanlip replied:
Oh, I think I misunderstood the point of this constant. DEFAULT_STRING_CLASSIFIER_ID should work.

seanlip replied:
(or maybe you should have a mapping from interaction ids to classifier ids?)

anmolshkl replied:
That's a good point! What should we call it? I think INTERACTION_CLASSIFIER_MAPPING sounds fine ...

anmolshkl replied:
Done.

seanlip commented on core/controllers/reader.py:
```diff
@@ -108,7 +110,7 @@ def classify_string_classifier_rule(state, normalized_answer):
         predicted_label = labels[0]
         if (predicted_label !=
-                lda_string_classifier.LDAStringClassifier.DEFAULT_LABEL):
+                sc.DEFAULT_LABEL):
```
Move to previous line and remove the parentheses.

Also, should this be a global concept, and not unique to the string classifier? If so, move DEFAULT_LABEL to feconf and have the string classifier refer to that.

anmolshkl replied:
I was going to ask you this, but then I thought one cannot keep putting all the constants in feconf, so I let it remain in StringlClassifier. I'll update it.

anmolshkl replied:
Done! I have named the constant as DEFAULT_CLASSIFIER_LABEL, sounds fine?

seanlip replied:
Should be OK!

seanlip commented on feconf.py:
```diff
@@ -348,6 +352,16 @@ def get_empty_ratings():
+# 'StringClassifiers' or 'CodeClassifiers' and the IDs of the
+# corresponding classifiers.
+ANSWER_CLASSIFIER_CLASSES = [{
+    'category': 'StringClassifiers',
```
You may not need a category here -- if it's here just for parity with interactions, you can remove it. For interactions, the category is needed to arrange them in the modal that pops up when the creator is asked to select an interaction.

anmolshkl replied:
The category here refers to StringClassifiers and CodeClassifiers. I think at some point this can help us to quickly know what are the StringClassifier and CodeClassifier algorithms available as opposed to putting everything in just one list.

BenHenning replied:
We can always retroactively refactor this list to include categories if we need them and they end up being useful (simply organizing it in a meaningful way with comments may be adequate to satisfy your concerns without the extra data structure complexity).

anmolshkl replied:
Done!

seanlip commented on core/tests/performance_tests/lda_string_classifier_performance_test.py:
```diff
@@ -23,9 +23,12 @@
 
 import os
 import time
+
```
Remove newline.

anmolshkl replied:
Done!

seanlip commented on core/tests/performance_tests/lda_string_classifier_performance_test.py:
```diff
@@ -23,9 +23,12 @@
 
-from core.domain.classifier.lda_string_classifier import LDAStringClassifier
+import feconf
+from core.domain.classifier import classifier_registry
```
This should go before the feconf import.

anmolshkl replied:
Done!

seanlip commented on core/domain/classifier/classifier_registry.py:
```diff
@@ -0,0 +1,97 @@
+import itertools
+import os
+import pkgutil
+import feconf
```
Add blank line above this one to separate it from system libs.

anmolshkl replied:
Done!

seanlip commented on core/domain/classifier/LDAStringClassifierTest.py:
```diff
@@ -14,8 +14,9 @@
 # limitations under the License.
 
-from core.domain.classifier.lda_string_classifier import LDAStringClassifier
+import feconf
```
Should go below "from core ..." imports

anmolshkl replied:
Done!

seanlip commented on core/controllers/reader_test.py:
```diff
@@ -135,13 +135,15 @@ def _init_classify_inputs(self, exploration_id):
+            feconf.DEFAULT_STRING_CLASSIFIER)
         string_classifier_predict = (
-            lda_string_classifier.LDAStringClassifier.predict)
+            sc.__class__.predict)
```
This seems a bit icky, but I noticed elsewhere (lda_string_classifier_performance_test.py) that you don't actually need to do `__class__`. So -- do you need it here?

anmolshkl replied:
Hmm, actually `lda_string_classifier.LDAStringClassifier` is doing something that we are not supposed to do and this is the reason why we have the registry in first place. So, though it looks icky but that was the only alternative I could think of. Oh, and I'm not sure how this works here (I'm guessing it needs a reference to the method to track how many times it is called, something which is not required in lda_string_classifier_performance_test). So the usecase is entirely different from lda_string_classifier_performance_test.py.

BenHenning replied:
Isn't predict an instance method? I would actually expect the swap on ``sc.predict`` to work here. What's the error if you do it that way?

anmolshkl replied:
That gives an error - TypeError: predict() takes exactly 2 arguments (3 given).

BenHenning commented:
LGTM once Sean's comments are resolved.

anmolshkl commented:
@seanlip I have addressed most of the issues. The only unresolved issue is - deciding an appropriate name for 'DEFAULT_STRING_CLASSIFIER' (which I think is ok). I'll be on vacation from 25-29 and would not be able to work on anything. If there's any change required, do let me know by today (your time). Thanks!

seanlip commented on core/controllers/reader.py:
```diff
@@ -20,8 +20,10 @@
 
 import jinja2
 
+import feconf
```
This and the next line should go under "from core.domain". (Alphabetize.)

anmolshkl replied:
Ah, I always mess up with imports! Thanks!

anmolshkl replied:
Done!

seanlip commented on core/controllers/reader.py:
```diff
@@ -183,9 +184,9 @@ def _get_exploration_player_data(
         'is_private': rights_manager.is_exploration_private(
             exploration_id),
-        # Note that this overwrites the value in base.py.
+        # Note that this overwrites the value in base_classifier.py.
```
I don't think this and the next line have anything to do with classifiers. base.py is the correct thing here.

anmolshkl replied:
Oh, that's pycharm refactoring at work. I had renamed base.py -> base_classifier.py and pycharm blindly replaced all the references in uncommitted files (not its fault). Thanks for catching that!

anmolshkl replied:
Done!

seanlip commented:
Yep! Sgtm.

On Tuesday, January 24, 2017, Anmol <notifications@github.com> wrote:

> *@anmolshkl* commented on this pull request.
> ------------------------------
>
> In feconf.py <https://github.com/oppia/oppia/pull/2926>:
>
> > @@ -68,6 +69,9 @@
>  RULES_DESCRIPTIONS_FILE_PATH = os.path.join(
>      os.getcwd(), 'extensions', 'interactions', 'rules.json')
>
> +# The ID of the classifier to be used as a default string classifier.
> +DEFAULT_STRING_CLASSIFIER = 'LDAStringClassifier'
>
> That's a good point! What should we call it? I think
> INTERACTION_CLASSIFIER_MAPPING sounds fine ...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/oppia/oppia/pull/2926>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AKFeygsKRFiHf_j3hplu_TTbUydkvMmqks5rVsozgaJpZM4LjpU8>
> .
>

seanlip commented:
(And thanks, @anmolshkl! Have a great vacation :) )
